### ISSUE TITLE
Add "no value" and "unknown value" as possible values for a Wikidata statement

### ISSUE THREAD

**Issue by SannitaSSJ** [OpenRefine/OpenRefine#1713]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**
It would be great if there could be a possibility to put "no value" or "unknown value" as a value for statements in a Wikidata scheme.

**Describe the solution you'd like**
When editing the Wikidata scheme for exporting values, it would be nice to have an additional command to add a "no value" or "unknown value" option, much like the Wikidata one (see image below).

![immagine](https://user-images.githubusercontent.com/31443821/44308657-12d51880-a3ba-11e8-8ce7-3730676e9548.png)

**Describe alternatives you've considered**
none

**Additional context**
none

**Comment by wetneb** [comment]
Indeed! Adding this to the todo list.

https://www.wikidata.org/wiki/Wikidata:Tools/OpenRefine/Editing/FAQ#Can_I_add_unknown_value_/_no_value_claims?

**Comment by SannitaSSJ** [comment]
Fantastic, thanks for updating the FAQs. :)

**Comment by wetneb** [comment]
Olivier from packed.be would also be interested in this.

**Comment by lozanaross** [comment]
@wetneb  just checking in to see if there's been any development on this or if there are any plans to implement this soon. It would actually be really valuable for custom Wikibase integrations, because apparently unknown values are not supported by wikidata integrator either, so not many options to work with that without resorting to manual editing.

**Comment by wetneb** [comment]
We can try to attract a bit more attention to this issue, since it is not that hard.

**Comment by sjathin** [comment]
I would like to work on this. As I am new to OpenRefine and Wikidata. I went through the following tutorial to get an idea - https://www.wikidata.org/wiki/Wikidata:Tools/OpenRefine/Editing/Tutorials/Basic_editing 

I followed the above tutorial, during the edit wikibase scheme, I added an item, and a statement(see the following image) should "no value" or "unknown value" be present here?

![Screenshot 2021-03-01 at 15 22 57](https://user-images.githubusercontent.com/29340368/109510623-87689200-7aa2-11eb-97e0-93d5b15daab0.png)

And also it would be helpful if I can get to know how to test this? Or if there is any documentation available that can help me to test this.

**Comment by wetneb** [comment]
Yes this is the area where we would add support for this. Here are a few ideas.

The simplest approach would be to replicate Wikibase's UI in OpenRefine (see the screenshot above). The advantage of this is that users will find it very easy to use, since they are probably already used to Wikibase's UI (and that would make it consistent with the rest of our schema editing UI, which closely mimics Wikibase's UI). But the downside is that this would not let us add any dependency on the tabular data. For instance consider the following table:

| organization | director |
|------------------|------------|
| Warner Bros. | Barry Meyer |
| Cycling Without Age | Lars Christiansen |
| Canadian Anarchist Federation |      |

Perhaps we want to upload these relations to Wikibase and use a "no value" statement for the last row. In that case our UI needs to be a bit different from Wikibase's own UI, to let this aspect of the schema be row-dependent.

I think the simplest way to implement this would be to introduce some special keywords which could be used as cell value and evaluate to "no value" or "unknown value".

For instance the user would first modify their table to get something like this:
| organization | director |
|------------------|------------|
| Warner Bros. | Barry Meyer |
| Cycling Without Age | Lars Christiansen |
| Canadian Anarchist Federation | #WB_NO_VALUE |

Then they would build a schema as usual, drag and drop the "director" column as statement value and that would generate a no value claim when the special keyword is encountered (and similarly for "unknown value"). The downside of this is that if you actually want to add "#WB_NO_VALUE" as a string-valued statement in Wikibase, you cannot do it anymore.

I'd be happy to hear more suggestions of how to proceed on this :)

To test this I think you could first write some backend unit tests. In addition it would be great to have some frontend tests as well but I think it could be a big task since we haven't started testing the Wikibase extension yet on this side.

**Comment by lozanaross** [comment]
Thanks for the ideas, Antonin. As a user, I can confirm that this solution would be quite helpful, and I don't think the downside of not being able to add "#WB_NO_VALUE" as a string-valued statement in Wikibase is a going to be a big detriment down the line.

If it's helpful, I could try testing with the Wikibases I have access to, once there is a version of the extension I can pull and test with.

**Comment by thadguidry** [comment]
@wetneb An alternative (and more complex but also more powerful and unlimited) would be introducing the Cell Variable approach that Ettore had asked for in that other issue.  Essentially Cell Flags or Stars or whatever dreamed up...instead of only have Row level Flags or Stars.  Then you could use that special signal of the Cell Flag, etc. to determine what to do regarding the statements sent to Wikibase.

**Comment by Binita-tech** [comment]
I have worked on two good first issues. Now I would like to work on this issue, kindly assign it to me.

**Comment by antoine2711** [comment]
@Binita-tech : this issue is more involving. There is a proposition to make as to where you control that, either in the WD Schema or at the cell edit level.

At the WD Schema level, it doesn't give you much power, and at the cell edit level, it migth prove not so simple…

Regards, Antoine

**Comment by trnstlntk** [comment]
We need this for the [Wikimedia Commons integration](https://github.com/orgs/OpenRefine/projects/2) too.

One frequently-used modeling convention on Wikimedia Commons (for creators of files imported from external websites) is the following:

![image](https://user-images.githubusercontent.com/9881645/168418024-b9d3f341-e04a-4271-bed1-6f858ba916c8.png)
Where [Creator](https://www.wikidata.org/wiki/Special:EntityPage/P170) is set to 'somevalue', with qualifiers. This is used extremely frequently and will be a common use case for file uploaders.

[Example file from Flickr](https://commons.wikimedia.org/wiki/File:PEN-dorp_(25846999062).jpg).

I was in the process of batch adding Creator statements with similar modeling, to [around 40,000 files on Wikimedia Commons that have been uploaded from iNaturalist](https://commons.wikimedia.org/wiki/Category:INaturalist_images_missing_SDC_depicts), but am currently unable to do so.

Adding this to the 'on the radar' tasks for the Commons integration!

**Comment by trnstlntk** [comment]
Terminology / naming related remark... in the UI on Wikimedia Commons the potential values are called `Custom value`/ `Some value` / `No value`.

![image](https://user-images.githubusercontent.com/9881645/168418307-2a139adb-2079-4ef6-9c55-d7eb6dbcaa7e.png)

**Comment by Abbe98** [comment]
@trnstlntk there is a [work in progress branch over at my fork](https://github.com/Abbe98/OpenRefine/pull/24/files). 

Technical issues aside one major question is how to make the columns with the special values droppable in the schema builder. Currently we do type checking against WB properties to allow or not allow certain columns, however, these special columns(which are just strings) should always be droppable. My local solution so far is to remove type checking all together.

**Comment by trnstlntk** [comment]
> Technical issues aside one major question is how to make the columns with the special values droppable in the schema builder. Currently we do type checking against WB properties to allow or not allow certain columns, however, these special columns(which are just strings) should always be droppable. My local solution so far is to remove type checking all together.

Mmmh @wetneb any opinions or ideas?

**Comment by Abbe98** [comment]
> Note that if you are dealing with a column of Wikibase entities, the other cells will still be reconciled, so the column metadata will still contain a reconciliation configuration as usual. This will make it possible to drag and drop this column in the corresponding input field as usual.

Should this be how it behaves currently?

**Comment by wetneb** [comment]
Yes, at the moment, the only sort of type checking we are doing on the columns is whether they are reconciled or not (and to the correct identifier space), for input fields which expect Wikibase entities. For all other fields, any column is accepted (since we do not have any column metadata which can give us insights about the type of values stored in them). If a column is partly reconciled, it will be usable in an input field for Wikibase entities, and all cells which are not reconciled will be ignored.

**Comment by Abbe98** [comment]
I think we might talk about different things because I do not recognize
that behavior. I'm only concerned about the UX/frontend logic for making
fields droppable(in schema-aligment.js). I currently need to comment out
the type checking part of that file to test the backend(which works as you
describe).

**Comment by wetneb** [comment]
Ok, then indeed I do not know what type checking you have in mind, it will probably be clearer once the front-end part of your changes are visible in your PR.

**Comment by trnstlntk** [comment]
I was wondering if finishing this request could be in scope for the Wikimedia 2023 grant. I notice I'm hesitant to document and teach people about the partial approach using #SOMEVALUE# and #NOVALUE# in the grid because, well, it's partial ;-)

**Comment by trnstlntk** [comment]
I have heard many questions about this, I have noticed that there's not much awareness of the partial solution now, and if folks know that it's possible, they can't find documentation on how to currently do some value (unknown value) / no value  statements in Wikibases.

To hopefully make this partial approach more visible and the approach discoverable, here's a demo video I just made.

https://commons.wikimedia.org/wiki/File:OpenRefine_-_somevalue_and_novalue_or_unknown_value_and_no_value.webm

**Comment by robertgarrigos** [comment]
Thanks, @trnstlntk, I was looking for this. I hope this issue could be finished someday. Until then I will use your solution.

**Comment by tfmorris** [comment]
@Devansh-18155 This is one of the issues we were discussing.

**Comment by Devansh-18155** [comment]
Sure @tfmorris, I will research about the issue soon. Thanks for helping and guiding me!

**Comment by Devansh-18155** [comment]
For this issue we can add a dropdown menu next to each value of inputs given by the user in Wikidata . This will allow users to switch between commands like "custom value", "no value", and "unknown value". Also for backend part we can implement two different classes one for "no value" and the other for "some value" this will enable the users to export and upload.

@tfmorris This is my implementation for the issue , Please let me know if you have any suggestions or i can go with this approach.

**Comment by Abbe98** [comment]
@Devansh-18155 this is already implemented on the backend and with magic values the only thing missing is the UI and docs. Ot might make sense to actually document the existing capability before attempting to expand it.

**Comment by Devansh-18155** [comment]
So the backend part is implemented ,then i will work on UI and docs integration for the schema builder
